src/roiextractors/extractors/tiffimagingextractors/scanimagetiffimagingextractor.py

```diff
@@ -22,13 +24,487 @@ def _get_scanimage_reader() -> type:
+            for i, imaging_extractor in enumerate(self._imaging_extractors):
+                frames[..., i] = imaging_extractor.get_frames(frame_idxs)
+        else:
+            return self.get_video(start_frame=frame_idxs[0], end_frame=frame_idxs[-1] + 1)
+
+    def get_image_size(self) -> Tuple:
+        """Get the size of a single frame.
+
+        Returns
+        -------
+        image_size: tuple
+            The size of a single frame (num_rows, num_columns, num_planes).
+        """
+        image_size = (*self._imaging_extractors[0].get_image_size(), self.get_num_planes())
+        return image_size
+
+    def get_num_planes(self) -> int:
+        """Get the number of depth planes.
+
+        Returns
+        -------
+        _num_planes: int
+            The number of depth planes.
+        """
+        return self._num_planes
+
+    def get_num_frames(self) -> int:
+        return self._imaging_extractors[0].get_num_frames()
+
+    def get_sampling_frequency(self) -> float:
+        return self._imaging_extractors[0].get_sampling_frequency()
+
+    def get_channel_names(self) -> list:
+        return self._imaging_extractors[0].get_channel_names()
+
+    def get_num_channels(self) -> int:
+        return self._imaging_extractors[0].get_num_channels()
+
+    def get_dtype(self) -> DtypeType:
+        return self._imaging_extractors[0].get_dtype()
+
+
+class ScanImageTiffMultiPlaneImagingExtractor(MultiPlaneImagingExtractor):
+    """Specialized extractor for reading multi-plane (volumetric) TIFF files produced via ScanImage."""
+
+    extractor_name = "ScanImageTiffMultiPlaneImaging"
+    is_writable = True
+    mode = "file"
+
+    def __init__(
+        self,
+        file_path: PathType,
+        channel: Optional[int] = 0,
+    ) -> None:
+        self.file_path = Path(file_path)
+        self.channel = channel
+        self.metadata = extract_extra_metadata(file_path)
+        parsed_metadata = parse_metadata(self.metadata)
+        num_planes = parsed_metadata["num_planes"]
+        imaging_extractors = []
+        for plane in range(num_planes):
+            imaging_extractor = ScanImageTiffSinglePlaneImagingExtractor(
+                file_path=file_path, channel=channel, plane=plane
+            )
+            imaging_extractors.append(imaging_extractor)
+        super().__init__(imaging_extractors=imaging_extractors)
+        assert all(
+            imaging_extractor.get_num_planes() == self._num_planes for imaging_extractor in imaging_extractors
+        ), "All imaging extractors must have the same number of planes."
+
+
+class ScanImageTiffSinglePlaneImagingExtractor(ImagingExtractor):
     """Specialized extractor for reading TIFF files produced via ScanImage."""
 
     extractor_name = "ScanImageTiffImaging"
     is_writable = True
     mode = "file"
 
+    def __init__(
+        self,
+        file_path: PathType,
+        channel: Optional[int] = 0,
+        plane: Optional[int] = 0,
```

Comment:
What is supposed to happen when `channel` or `plane` is `None`? It looks like it would cause an error at the inequality assertions below

Reply:
Also as per discussion in #245, I'm in the group that thinks these ought to be strings to be as flexible and universal as possible

I know for certain formats, the `channel` name is entirely determined by the user through the acquisition interface and it doesn't need any ordinality or even a number in it at all (especially common to see them called simply by the approximate color they filter, e.g., 'blue_channel', 'green_channel', etc.

Reply:
> What is supposed to happen when channel or plane is None? It looks like it would cause an error at the inequality assertions below

`plane` and `channel` default to `0` -- they should never be `None`.

> Also as per discussion in https://github.com/catalystneuro/roiextractors/issues/245, I'm in the group that thinks these ought to be strings to be as flexible and universal as possible
> 
> I know for certain formats, the channel name is entirely determined by the user through the acquisition interface and it doesn't need any ordinality or even a number in it at all (especially common to see them called simply by the approximate color they filter, e.g., 'blue_channel', 'green_channel', etc.

Ok, then we can specify `channel_name` instead of `channel` (index).  I still think `plane` should be an index though, since for large numbers of depth planes there won't be a unique name for each plane.

Reply:
> since for large numbers of depth planes there won't be a unique name for each plane.

Why would there not be a unique name? And why would it not in principle be user specifiable on any given acquisition system? The only reason I'd see to keep it a numeric int would be if you wanted to assert something about the ordinality, but it also ought to be possible to specify an acquisition order that is not contiguous over a Z direction

Maybe @alessandratrapani can shed light on this; for something like Bruker, can a user design/specify a cycle or alternation of cycles that are not strictly increasing/decreasing with respect the Z plane?

Reply:
I think that Paul claim here is really reasonable and your demands too strong. You are a kind of asking him prove a negative (are YOU really sure that no system could decide to use names for their frames?). Can you elaborate on where are you coming from? I really have a hard time imaging how it would make sense to put names to planes.

Even if the numbers are not ordered (cycles mostly are though) numbers are still seem like the best way to refer to planes. You really need to have meanigful names for strings to make sense and if that is the case we should use the concept of stream to refer to it.

Reply:
> Why would there not be a unique name?

I just mean that the plane wouldn't be an inherently *named* quantity in the way that channels are ('red channel', 'green channel', etc.).  But the planes are an inherently *ordered* quantity (plane 0, plane 1, ..., plane N from the bottom/top).

Reply:
> I really have a hard time imaging how it would make sense to put names to planes.

`_planeA_channelRed`
`_planeB_channelBlue`

`_cycle_{x}coord_{y}coord_{z}coord`

`_{uuid()}_chan0` (hash for each plane); e.g., `_a937315b-2c1b-4cd7-abfb-d5efc562deac_chan0`

I'm saying the fact I _can_ imagine these cases doesn't mean any microscope company ever _would_ or even _has_, which is why I appeal to someone with extensive experience to definitely say 'every microscope company in the world would save plane ID information as an integer?'. That seems too specific a claim to possibly be true, so why should we assume it here?

Whereas 'can every plane ID generated with any device be mapped as a string?' seems much safer to me. Also tautological in the case where the information is available in the file names (whose stems are string) or the metadata stored in a text-based header, because that's literally how it had to have gotten transcoded as bytes anyhow. And we already have cases of that seen witth Bruker



> I think that Paul claim here is really reasonable and your demands too strong.

Sorry if I phrased too broad of a question

I'm only after the most generic solution for a universal API that will not require us to go back and re-implement this when we see it occur in a future format, and to me even if you do nothing more than `int_as_str = str(input_you_expect_to_be_an_int)` in this case I'd be happy; e.g., `plane: str = "0"`.

No matter what, you can always map `int` -> `str` easily, and the latter is much more extensible. The only thing you lose is perhaps ordinal comparison, but that's why I also wonder if it's always necessary that the order in which cycles are captured is necessarily related to another ordinal property such as space (or can planes go back and forth non-linearly?)



> Even if the numbers are not ordered (cycles mostly are though) numbers still seem like the best way to refer to planes (see the cyclic group common representation).

> If give you my format, you give me an object that allows me to extract my data ergonomically and efficiently with a common API. I don't need to know how you do it.

As you say, the _how_ of implementing plane identification and referencing within our code might very well map/order the available plane selections onto a cyclic group since that is how the device would some way or another go about capturing them, and regardless of the identifiers of those planes we can always map them to a finite set of integer indices internally

But as you also say,  the users don't need to know how _we_ do it under the hood.

Let's say a user acquiring multiple planes does not care about this implicit order - why would we force them, via the top level usage of our common API, to know what that integer-sensitive indexing is?

As opposed to just fetching for them the available, unordered set of string options, and allowing them to select the one they want to extract?

Also, if you feel strongly in favor of duck typing, you could always make everyone happy and do both 😄

Reply:
Actually, there's already a practical case and we don't have to speculate about what might be possible.

Multi-plane support for reading from NWB files

A good user following best practices might apply an inferrable ordinal labeling of each imaging plane, e.g., `ImagingPlane0`, `ImagingPlane1`, etc.

Or they might have given it something richer and more identifiable.

`ImagingPlaneCortex`

`ImagingPlaneV1`

`ImagingPlaneHippocampus`

Sure we could arbitrarily cast these together into a list (order probably determined by the order in which they were written to the file) and reference to each element by their index in that list; but it wouldn't be as intuitive, or echo that richness of annotation.

Reply:
Moving this to a dedicated PR #248

Reply:
> I'm only after the most generic solution for a universal API that will not require us to go back and re-implement this when we see it occur in a future format, and to me even if you do nothing more than int_as_str = str(input_you_expect_to_be_an_int) in this case I'd be happy; e.g., plane: str = "0".

Ok, I think this is reasonable.  For ScanImage (and I suspect for a lot of formats) using an integer for each plane is the most parsimonious solution.  But, as you pointed out, there may be some cases where strings are more appropriate esp. when analyzing depth planes separately rather than in a volumetric fashion.  More importantly, it is very easy to take `plane_name = '0'` instead of `plane=0` but much harder to go the other way.

Reply:
Those are some good points, @CodyCBakerPhD, thanks for elaborating as I requested. Specially the example of NWB is very illustrative. I will integrate this in my reasoning about https://github.com/catalystneuro/roiextractors/issues/245.

Spikeinterface does have int | str type for channel_ids and I think that that bothers me more than overgenerality.

Reply:
> Maybe @alessandratrapani can shed light on this; for something like Bruker, can a user design/specify a cycle or alternation of cycles that are not strictly increasing/decreasing with respect the Z plane?
I am sorry for coming late for this discussion. 
I was reasoning on edge cases where the plane acquisition is not strictly increasing/decreasing along depth, but I couldn't see how and why it should happen. 
However, this stands just for volumetric imaging acquisition, where you want to record the activity of a specific brain area without caring (a priori) how the neurons are distributed in the volume. 
Imagine the case where you want to record pyramidal cells in a cortical column, but you want to do it in a specific order that follows functional constraints along the layers distribution, not necessarily depth. In this very specific case, you could define a cycle acquisition that acquires planes following a functional order. Therefore, you might also want to name the plane according to the specific region/layer they are referred to. (Note that this is just a speculation of what might come in the future; I have never encountered such cases.)
This is why I endorse the solution of defining `plane_name` as a string that can be just an int_as_str for most of the cases.